dsl/src/main/scala/com.crobox.clickhouse/dsl/AggregationFunctions.scala

```diff
@@ -0,0 +1,259 @@
+package com.crobox.clickhouse.dsl
+
+import com.crobox.clickhouse.dsl.AggregateFunction.Combinator
+import com.crobox.clickhouse.dsl.AnyResult.{AnyModifier, AnyResultDsl}
+import com.crobox.clickhouse.dsl.Leveled.{LevelModifier, LevelModifierDsl, LeveledAggregatedFunction}
+import com.crobox.clickhouse.dsl.Sum.{SumDsl, SumModifier}
+import com.crobox.clickhouse.dsl.TableColumn.AnyTableColumn
+import com.crobox.clickhouse.dsl.Uniq.{UniqDsl, UniqModifier}
+import com.crobox.clickhouse.time.MultiInterval
+import org.joda.time.DateTime
+
+//https://clickhouse.yandex/docs/en/agg_functions/reference
+abstract class AggregateFunction[V](targetColumn: AnyTableColumn) extends ExpressionColumn[V](targetColumn)
+
+case class CombinedAggregatedFunction[T <: TableColumn[_], Res](combinator: Combinator[T, Res],
+                                                                target: AggregateFunction[_])
+    extends AggregateFunction[Res](EmptyColumn())
+
+object AggregateFunction {
+  type StateResult = String
+  sealed trait Combinator[T <: Column, Result]
+
+  case class If[T <: Column, Res](condition: Comparison)      extends Combinator[T, Res]
+  case class CombinatorArray[T <: TableColumn[Seq[V]], V]()   extends Combinator[T, V]
+  case class ArrayForEach[T <: TableColumn[Seq[Seq[V]]], V]() extends Combinator[T, Seq[V]]
+  case class State[T <: Column]()                             extends Combinator[T, StateResult]
+  case class Merge[T <: TableColumn[StateResult], Res]()      extends Combinator[T, Res]
+
+  trait AggregationFunctionsCombinersDsl {
+
+    def aggIf[T <: TableColumn[Res], Res](condition: Comparison)(aggregated: AggregateFunction[Res]) =
```

Comment:
Here as well as on other functions wich can be composed out of parts you might consider a flow to work with that more resembles the original query.

Its getting pretty obscure with the current way of defining uniqIf statements for instance, while it could look like this: 

`uniq(if(notEmpty(session_id), MyAggCol)))`

Reply:
The choice for the state combinators is connected to the fact that `if,state,merge` are global available for all the aggregation functions, and the clickhouse system for it is so flexible that it will be impossible to explicitly declare all the support combinations.
Take for example the function `uniqIfIfIfIfIfArrayState`, this would be a valid aggregation function which would have to have 5 if conditions.